src/Client/HedgedConnectionsFactory.cpp

```diff
@@ -0,0 +1,360 @@
+#if defined(OS_LINUX)
+
+#include <Client/HedgedConnectionsFactory.h>
+#include <Common/typeid_cast.h>
+
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int ALL_CONNECTION_TRIES_FAILED;
+    extern const int ALL_REPLICAS_ARE_STALE;
+}
+
+HedgedConnectionsFactory::HedgedConnectionsFactory(
+    const ConnectionPoolWithFailoverPtr & pool_,
+    const Settings * settings_,
+    const ConnectionTimeouts & timeouts_,
+    std::shared_ptr<QualifiedTableName> table_to_check_)
+    : pool(pool_), settings(settings_), timeouts(timeouts_), table_to_check(table_to_check_), log(&Poco::Logger::get("HedgedConnectionsFactory"))
+{
+    shuffled_pools = pool->getShuffledPools(settings);
+    for (size_t i = 0; i != shuffled_pools.size(); ++i)
+        replicas.emplace_back(ConnectionEstablisher(shuffled_pools[i].pool, &timeouts, settings, table_to_check.get()));
+
+    max_tries
+        = (settings ? size_t{settings->connections_with_failover_max_tries} : size_t{DBMS_CONNECTION_POOL_WITH_FAILOVER_DEFAULT_MAX_TRIES});
+
+    fallback_to_stale_replicas = settings && settings->fallback_to_stale_replicas_for_distributed_queries;
+    entries_count = 0;
+    usable_count = 0;
+    failed_pools_count = 0;
+}
+
+HedgedConnectionsFactory::~HedgedConnectionsFactory()
+{
+    pool->updateSharedError(shuffled_pools);
+}
+
+std::vector<Connection *> HedgedConnectionsFactory::getManyConnections(PoolMode pool_mode)
+{
+    size_t min_entries = (settings && settings->skip_unavailable_shards) ? 0 : 1;
+
+    size_t max_entries;
+    if (pool_mode == PoolMode::GET_ALL)
+    {
+        min_entries = shuffled_pools.size();
+        max_entries = shuffled_pools.size();
+    }
+    else if (pool_mode == PoolMode::GET_ONE)
+        max_entries = 1;
+    else if (pool_mode == PoolMode::GET_MANY)
+        max_entries = settings ? size_t(settings->max_parallel_replicas) : 1;
+    else
+        throw DB::Exception("Unknown pool allocation mode", DB::ErrorCodes::LOGICAL_ERROR);
+
+    std::vector<Connection *> connections;
+    connections.reserve(max_entries);
+    Connection * connection = nullptr;
+
+    /// Try to start establishing connections with max_entries replicas.
+    for (size_t i = 0; i != max_entries; ++i)
+    {
+        int index = startEstablishingNewConnection(connection);
+        if (index == -1)
+            break;
+        if (replicas[index].is_ready)
+            connections.push_back(connection);
+    }
+
+    /// Process connections until we get enough READY connections
+    /// (work asynchronously with all connections we started).
```

Comment:
This is pretty cool, and it is async indeed (because we start all connections in a loop before).
But it is not obvious, especially that we use `blocking` flag for `getNextConnection`.
Probably, we just need some better comment here.
Or we can explicitly call `processEpollEvents(/*blocking*/ = true, connection_);` and `setBestUsableReplica()` (cause `getNextConnection` is simple itself).

Reply:
BTW, technically it is possible to start reading packets from ready connection as soon as we got it, not even waiting for other.
But, we will not know total number of offsets.
But, in case of GET_ALL `min_entries = max_entries`, so we actually know it :). If any connection fails, we can cancel query for the others.
Maybe we should TODO it (or just add comment).